metricbeat/module/prometheus/remote_write/_meta/docs.asciidoc

```diff
@@ -7,6 +7,18 @@ remote_write:
   - url: "http://localhost:9201/write"
 ------------------------------------------------------------------------------
 
+TIP: It is recommended to carefully choose the https://prometheus.io/docs/practices/remote_write/#parameters[parameters]
```

Comment:
are there other recommendations we can make? if not, i would make it even more focused on `max_shards` instead talking about general tuning - which i think can be very difficult for the user to do.

Additionally, I would try to give some recommendation around `max_shards`. For example, if you have one metricbeat, which value makes sense for `max_shards`? If you have X metricbeats, which value then makes sense?

Reply:
Hey @roncohen, thanks for commenting here! Answering inline:

> are there other recommendations we can make? if not, i would make it even more focused on `max_shards` instead talking about general tuning - which i think can be very difficult for the user to do.
> 

I think that we already focusing on `max_shards` in the biggest part of this section. `max_shards` option is explained in detail and gives the overview of under which circumstances it should be set to something other than the default value.

> Additionally, I would try to give some recommendation around `max_shards`. For example, if you have one metricbeat, which value makes sense for `max_shards`? If you have X metricbeats, which value then makes sense?

`max_shards` is an option that is set for each remote_write `url` (endpoint) explicitly. So in case we have more than one `url`s, and hence metricbeats, each metricbeat will have its own configuration which of course can have the same value for all of them. 

As far as the recommended values is concerned, I think that we cannot provide something so specific here. From what I've seen so far while testing and running some basic benchmarks on it, it depends. For instance you can have one Metricbeat and this to be enough for a Prometheus server with 3K timeseries, but what if we have a Prometheus with 200K timeseries? The number of timeseries is key factor that will make the one instance of metricbeat to potentially fall behind. In addition, this setup will behave differently in case of having prometheus scraping interval equal to `5s` (will push the limits) and in case of having sth like `5m` where the load will not be constant.

Let me know what you think!

Reply:
let me ask differently. When does it makes sense to have a `max_shards` > 1?

Reply:
btw. to scale metricbeats, wouldn't you instead use one url and then put metricbeat behind a loadbalancer?

Reply:
Metricbeat needs some time to convert incoming metrics and putting them into the output queue. If we set `max_sharts > 1` we can run this process concurrently, probably gaining in overall throughput.

From what I see by default each request will send up to 100 samples (`max_samples_per_send`). And we do some work to compact them in as few documents as possible. So each request will result in << 100 docs to send to Elasticsearch.

I would expect Elasticsearch to always be the limiting factor here, but it would be worth checking that we can keep the queue as its max throughput with a `max_shards` as low as 1. If not, I would expect ~10 to be a sane default for a single Metricbeat instance.

Reply:
Also, our queue `min_events` setting is set to 512 by default: https://www.elastic.co/guide/en/beats/filebeat/master/configuring-internal-queue.html#_flush_min_events. So that means we should benefit from having more shards filling that concurrently (as opposed as a single thread sending <100 docs per request).

Reply:
so it sounds like `max_shards` is useful to increase concurrency of a single metricbeat instance. In that case, it might make sense to set `max_shards` to the number of cores of the fleet of metricbeats? That assumes metricbeat processing actually scales with the number of cores. 

Separately, I'd be curious to know if it might make sense to increase the `max_samples_per_send` to decrease the relative overhead of sending a batch to increase overall throughput. Would be great to have more data on this.

Hopefully you see my point: We, the developers, have a somewhat hard time reasoning about how the values interact, so it's a bit cheeky to ask users to "carefully choose the parameters".

Reply:
Yes, I definitely see the point of gaining & providing more understanding about how these parameters affect the performance. @ChrsMark it would be interesting to dig a bit deeper and test a little more maybe? (You should not have told me you like benchmarking :stuck_out_tongue:)

A few thoughts on what we should pay attention to, as these are the variables we want to explain to users. This is what I believe the overall behavior will be, but we would need to confirm these hypothesis.

- Increasing `max_shards` setting (over 1) should increase the overall throughput (written docs/sec), until things saturate and increasing it doesn't have any benefit. I would expect this golden number to be around the number of cores of the machine (as @roncohen said). Increasing this setting will also increase MB mem & CPU usage.

- Increasing `max_samples_per_send` should increase the overall throughput, and it should saturate when the resulting document count per send reaches [bulk_max_size](https://www.elastic.co/guide/en/beats/filebeat/current/elasticsearch-output.html#_bulk_max_size) (50 by default, for the ES output). I realize now this is more important than `min_events`, as that one should only affect at the start time. Increasing this setting will also increase MB mem usage.

These 2 settings will affect MB memory usage, which is an important factor: `max_shards x max_smaples_per_send x memory used per send`

All in all I would expect max_shards == number of MB cores to be a good default, leaving the rest untouched. Always minding their relation with the Metricbeat output queue settings, and the chosen output.

Reply:
Well I got all these points. Something important here is that if Prometheus stores a lot of timeseries and it collects them in a frequent rate (like 15sec for instance) then all these metrics will be queued in Prometheus queue so as to be sent in bathces of `100`. So if the number of metrics is 100K then we have 100K/100=1K batches to send  sequentially in a time window of 15 seconds because after that period we will receive another group of 1K batches. In this case, I guess that Prometheus would try to increase the number of shards.

With this example I want to emphasise that Prometheus side according to each specific case (number of timeseries and scrape interval) has a significant role here which we cannot predict on our side, and this makes the problem even more multidimensional. 

Regarding `max_samples_per_send`. If we increase this then Metricbeat would have more metrics to process and in case of having only one shard then we will have to consume less batches but each batch would take more to be consumed so it might be the somehow the same.

btw @roncohen there are some benchmarks in https://github.com/elastic/beats/issues/17079. Nothing special but we could revisit this benchmarking story with all these we discuss here and perform benchmarks with big datasets and for different values in all these dimensions.

Reply:
Hey folks! I run some extra tests (see https://github.com/elastic/beats/issues/17079#issuecomment-612878489) and it is true that `max_shards`' golden number seems to be the number of cores. On a 8-core machine it would be 8 (practically I set 6: 6 cores for handling the remote connections, 1 for main routine and leave 1 free for the system).

`max_samples_per_send` can give slightly better throughput (not something extremely important).

@roncohen @exekias find more details in https://github.com/elastic/beats/issues/17079#issuecomment-612878489 and as far as the documentation is concerned I think that we could suggest directly those 2 options:

- `max_shards`: maximum the number of cores. Metricbeat can handle remote storage connections in parallel but increasing this to more than the number of cores is not suggested.
- `max_samples_per_send`: Metricbeat can handle big chunks of metrics (500, 1000) with a small increase in the throughput. However this will result in more memory usage for Metricbeat.